lib/hyacinth/batch_import/index_prerequisite_map.rb

```diff
@@ -0,0 +1,237 @@
+
+        row_numbers_to_parent_row_numbers = {}
+        row_numbers_to_parent_identifiers.each do |row_number, parent_identifiers|
+          # Remember: This row may have multiple parents, so that's why parent_row_numbers, below,
+          # is an array rather than a single value.
+          parent_row_numbers = []
+          parent_identifiers_not_in_csv = []
+          parent_identifiers.each do |identifier|
+            if identifiers_to_row_numbers.key?(identifier)
+              parent_row_numbers << identifiers_to_row_numbers[identifier]
+            else
+              parent_identifiers_not_in_csv << identifier
+            end
+          end
+
+          # For any parent identifiers not found in the csv, let's check if they're referring to
+          # existing digital object in Hyacinth.
+          unresolvable_identifiers = []
+          parent_identifiers_not_in_csv.each do |identifier|
+            found_object_uids = Hyacinth::Config.digital_object_search_adapter.identifier_to_uids(identifier)
+            if found_object_uids.blank?
+              # If we found zero uids for the given identifier, that's not good because that
+              # means the csv data refers to a parent that doesn't exist in the spreadsheet
+              # and also doesn't exist in Hyacinth. We'll log this unresolvable identifier
+              # and later on we'll raise an error.
+              unresolvable_identifiers << identifier
+              next
+            elsif found_object_uids.length > 1
+              # If we found more than one uid for the given identifier, that's not good because that
+              # means we have an ambiguous parent reference. We'll raise an error.
+              raise Hyacinth::Exceptions::BatchImportError,
+                    "Ambiguous parent reference. Parent identifier #{identifier} was resolved to "\
+                    "#{found_object_uids.length} objects in Hyacinth (#{found_object_uids.join(', ')}), "\
+                    "so it's unclear which parent is being referenced."
+            end
+
+            # If we got here, that means we found just one uid for the given identifier. That's great.
+            # We'll use that UID as the "parent row number" value in the map, and it will be treated
+            # in a special way later on, differently than other numeric values.
+            parent_row_numbers << identifiers_to_row_numbers[found_object_uids.first]
+          end
+
+          # Raise an error if we encountered any unresolvable identifiers
+          if unresolvable_identifiers.present?
+            raise Hyacinth::Exceptions::BatchImportError,
+                  "The following parent identifiers could not be found in CSV data or "\
+                  "among existing Hyacinth objects: #{unresolvable_identifiers.join(', ')}"
+          end
+
+          # If there are any duplicate values in parent_row_numbers, that indicates that the
```

Comment:
The parent_uids attribure on DigitalObject::Base is a Set, so it shouldn't make it into our data model.  But also, if we encounter a duplicate parent reference in a user-submitted CSV then it's definitely a mistake, and data that we don't want to send downstream to other prerequisite-related operations, so that's why I'm raising the error here.

It's specifically about this kind of case:
_parent_digital_objects[0].identifier = 'parent1'
_parent_digital_objects[1].identifier = 'parent2'
_parent_digital_objects[2].identifier = 'parent1'

And this (probably extremely rare) other case:
_parent_digital_objects[0].uid = 'uid1'
_parent_digital_objects[1].uid = 'uid2'
_parent_digital_objects[2].identifier = 'identifier-that-points-to-record-with-uid1'

Reply:
OHH okay, I totally did not understand what this check was doing. Thanks for explaining!